featurebyte/query_graph/sql/aggregator/forward.py

```diff
@@ -0,0 +1,155 @@
+        specifications provided can be done in a single groupby operation.
+
+        Parameters
+        ----------
+        specs: list[ForwardAggregateSpec]
+            Aggregation specifications
+
+        Returns
+        -------
+        LeftJoinableSubquery
+        """
+        spec = specs[0]
+
+        # End point expression
+        point_in_time_expr = get_qualified_column_identifier(SpecialColumnName.POINT_IN_TIME, "REQ")
+        point_in_time_epoch_expr = self.adapter.to_epoch_seconds(point_in_time_expr)
+        horizon_in_seconds = 0
+        if spec.parameters.horizon:
+            horizon_in_seconds = pd.Timedelta(spec.parameters.horizon).total_seconds()
+        end_point_expr: Expression = cast(
+            Expression,
+            parse_one(f"FLOOR({point_in_time_epoch_expr.sql()} + {horizon_in_seconds})"),
+        )
+
+        # Get valid records (timestamp column is within the point in time, and point in time + horizon)
+        # TODO: update to range join
+        record_validity_condition = expressions.and_(
+            expressions.GT(
+                this=get_qualified_column_identifier(spec.parameters.timestamp_col, "TABLE"),
+                expression=point_in_time_expr,
+            ),
+            expressions.LTE(
+                this=get_qualified_column_identifier(spec.parameters.timestamp_col, "TABLE"),
+                expression=end_point_expr,
+            ),
+        )
+
+        # Join the valid records, and the request table
+        groupby_keys = [
+            GroupbyKey(
+                expr=point_in_time_expr,
+                name=SpecialColumnName.POINT_IN_TIME,
+            )
+        ] + [
+            GroupbyKey(
+                expr=get_qualified_column_identifier(serving_name, "REQ"),
+                name=serving_name,
+            )
+            for serving_name in spec.serving_names
+        ]
+        value_by = (
+            GroupbyKey(
+                expr=get_qualified_column_identifier(spec.parameters.value_by, "TABLE"),
+                name=spec.parameters.value_by,
+            )
+            if spec.parameters.value_by
+            else None
+        )
+        groupby_columns = [
+            GroupbyColumn(
+                agg_func=s.parameters.agg_func,
+                parent_expr=(
+                    get_qualified_column_identifier(s.parameters.parent, "TABLE")
+                    if s.parameters.parent
+                    else None
+                ),
+                result_name=s.agg_result_name,
+            )
+            for s in specs
+        ]
+        groupby_input_expr = (
+            select()
+            .from_(
+                expressions.Table(
+                    this=quoted_identifier(self.request_table_plan.get_request_table_name(spec)),
+                    alias="REQ",
+                )
+            )
+            .join(point_in_time_expr, join_type="inner", on=record_validity_condition)
```

Comment:
I think this should be joining with `spec.source_expr.subquery(alias="TABLE")` instead of `point_in_time_expr`. Using the `TABLE` alias here to match with the expressions defined above.